core/src/main/java/org/apache/calcite/rel/rules/JoinToCorrelateRule.java

```diff
@@ -56,31 +60,57 @@
  *
  * <p>would require emitting a NULL emp row if a certain department contained no
  * employees, and Correlator cannot do that.</p>
+ *
+ * <p>There is a constructor parameterized to allow any sub-class of
+ * {@link org.apache.calcite.rel.core.Join}, not just
+ * {@link org.apache.calcite.rel.logical.LogicalJoin}.</p>
  */
 public class JoinToCorrelateRule extends RelOptRule {
+
+  private final Function<Join, SemiJoinType> semiJoinTypeFunction;
+
   //~ Static fields/initializers ---------------------------------------------
 
   public static final JoinToCorrelateRule INSTANCE =
       new JoinToCorrelateRule(RelFactories.LOGICAL_BUILDER);
 
+  /**
+   * Rule that converts a {@link org.apache.calcite.rel.core.SemiJoin}
+   * into a {@link org.apache.calcite.rel.logical.LogicalCorrelate}
+   */
+  public static final JoinToCorrelateRule SEMI =
+      new JoinToCorrelateRule(SemiJoin.class,
+              RelFactories.LOGICAL_BUILDER, join -> SemiJoinType.SEMI);
+
   //~ Constructors -----------------------------------------------------------
 
   /**
    * Creates a JoinToCorrelateRule.
    */
   public JoinToCorrelateRule(RelBuilderFactory relBuilderFactory) {
-    super(operand(LogicalJoin.class, any()), relBuilderFactory, null);
+    this(LogicalJoin.class, relBuilderFactory, join -> SemiJoinType.of(join.getJoinType()));
   }
 
   @Deprecated // to be removed before 2.0
   protected JoinToCorrelateRule(RelFactories.FilterFactory filterFactory) {
     this(RelBuilder.proto(Contexts.of(filterFactory)));
   }
 
+  /**
+   * Creates a JoinToCorrelateRule for a certain sub-class of
+   * {@link org.apache.calcite.rel.core.Join}
+   */
+  public JoinToCorrelateRule(Class<? extends Join> clazz,
+                             RelBuilderFactory relBuilderFactory,
+                             Function<Join, SemiJoinType> semiJoinTypeFunction) {
+    super(operand(clazz, any()), relBuilderFactory, null);
```

Comment:
Leaving the description parameter as null is not a good idea. The rules used by the planner need to have a unique description. INSTANCE and SEMI above will end up with the description which means that they cannot be used together in the same planner.